File: onyx-database/src/main/java/com/onyx/util/EntityClassLoader.java
package com.onyx.util;

import com.onyx.entity.SystemAttribute;
import com.onyx.entity.SystemEntity;
import com.onyx.entity.SystemIndex;
import com.onyx.entity.SystemRelationship;
import com.onyx.exception.OnyxException;
import com.onyx.persistence.annotations.values.CascadePolicy;
import com.onyx.persistence.annotations.values.FetchPolicy;
import com.onyx.persistence.annotations.values.IdentifierGenerator;
import com.onyx.persistence.annotations.values.RelationshipType;
import com.onyx.persistence.context.SchemaContext;
import com.onyx.persistence.context.impl.CacheSchemaContext;

import javax.tools.JavaCompiler;
import javax.tools.StandardJavaFileManager;
import javax.tools.StandardLocation;
import javax.tools.ToolProvider;
import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.io.Writer;
import java.lang.reflect.Method;
import java.net.URL;
import java.net.URLClassLoader;
import java.util.*;
import java.util.concurrent.atomic.AtomicReference;
import java.util.function.Consumer;

/**
 * Created by timothy.osborn on 3/6/15.
 * <p>
 * This class saves the entity information and formats the source on disk
 */
@SuppressWarnings("WeakerAccess")
public class EntityClassLoader {

    public static final Set<String> LOADED_CLASSES = new HashSet<>();
    public static final Set<String> WRITTEN_CLASSES = new HashSet<>();

    public static final String GENERATED_DIRECTORY = "generated";
    public static final String GENERATED_ENTITIES_DIRECTORY = GENERATED_DIRECTORY + File.separator + "entities";
    public static final String GENERATED_QUERIES_DIRECTORY = GENERATED_DIRECTORY + File.separator + "queries";

    public static final String SOURCE_DIRECTORY = "source";
    public static final String XTEND_SOURCE_DIRECTORY = "xtend";
    public static final String SOURCE_ENTITIES_DIRECTORY = SOURCE_DIRECTORY + File.separator + "entities";
    public static final String XTEND_SOURCE_ENTITIES_DIRECTORY = XTEND_SOURCE_DIRECTORY + File.separator + "entities";

    /**
     * Write an enum to file
     * @param enumClass Enum Class Name
     * @param enumValues Enum values separated by comman and ended with a semicolon.
     * @param outputDirectory Source directory
     */
    public static void writeEnum(String enumClass, String enumValues, String outputDirectory) {
        String[] nameTokens = enumClass.split("\\.");
        String packageName = "package ";
        for (int i = 0; i < nameTokens.length - 1; i++)
            packageName = packageName + nameTokens[i] + ".";

        packageName = packageName.substring(0, packageName.length() - 1);
        packageName = packageName + ";\n\n";

        String classDef = "public enum " + nameTokens[nameTokens.length - 1] + " { \n";
        String end = "\n}";
        String classContents = packageName + classDef + enumValues + end;

        writeClassContents(outputDirectory, enumClass, classContents);
    }

    /**
     * Helper method used to write the class contents to a file
     *
     * @param outputDirectory Source directory
     * @param className Name of class including package
     * @param classContents Class source code
     * @param extension file extension
     */
    @SuppressWarnings("ResultOfMethodCallIgnored")
    private static void writeClassContents(String outputDirectory, String className, String classContents, String extension) {
        try {
            // File output
            final File classFile = new File(outputDirectory + File.separator +
                    className.replaceAll("\\.", "/") + extension);
            classFile.getParentFile().mkdirs();
            classFile.createNewFile();

            final Writer file = new FileWriter(outputDirectory + File.separator +
                    className.replaceAll("\\.", "/") + extension);

            file.write(classContents);
            file.flush();

            file.flush();
            file.close();

        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    /**
     * Helper method used to write the class contents to a file
     *
     * @param outputDirectory Source directory
     * @param className Name of class including package
     * @param classContents Class source code
     */
    @SuppressWarnings("ResultOfMethodCallIgnored")
    private static void writeClassContents(String outputDirectory, String className, String classContents) {
        writeClassContents(outputDirectory, className, classContents, ".java");
    }

    public static void writeXtendClass(final SystemEntity systemEntity, final String databaseLocation, SchemaContext schemaContext, String idType, String idName) {
        final String packageName = systemEntity.getName().replace("." + systemEntity.getClassName(), "");
        final String outputDirectory = databaseLocation + File.separator + XTEND_SOURCE_ENTITIES_DIRECTORY;
        final String className = systemEntity.getClassName();
        final String generatorType = IdentifierGenerator.values()[systemEntity.getIdentifier().getGenerator()].toString();
        final String fileName = systemEntity.getFileName();
        final String idLoadFactor = String.valueOf(systemEntity.getIdentifier().getLoadFactor());

        final StringBuilder builder = new StringBuilder();
        builder.append("package ");
        builder.append(packageName);
        builder.append("\n" +
                "\n" +
                "import com.onyx.persistence.annotations.*\n" +
                "import com.onyx.persistence.*\n" +
                "import net.sagittarian.onyx.annotations.OnyxFields\n" +
                "import net.sagittarian.onyx.annotations.OnyxJoins\n" +
                "import org.eclipse.xtend.lib.annotations.Accessors" +
                "\n\n" +
                "@Entity(fileName = \"")
                .append(fileName)
                .append("\")\n")
                .append("@OnyxFields\n")
                .append("@OnyxJoins\n")
                .append("@Accessors\n")
                .append("class ")
                .append(className)
                .append(" extends ManagedEntity implements IManagedEntity\n{\n\n")
                .append("    @Attribute\n")
                .append("    @Identifier(generator = ")
                .append(generatorType)
                .append(", loadFactor = ")
                .append(idLoadFactor)
                .append(")\n    ")
                .append(idType)
                .append(" ")
                .append(idName)
                .append(";\n\n");

        for (final SystemAttribute attribute : systemEntity.getAttributes()) {
            if (attribute.getName().equals(systemEntity.getIdentifier().getName())) {
                continue;
            }

            if (attribute.isEnum()) {
                writeEnum(attribute.getDataType(), attribute.getEnumValues(), outputDirectory);
            }

            boolean isIndex = false;
            for (SystemIndex indexDescriptor : systemEntity.getIndexes()) {
                if (indexDescriptor.getName().equals(attribute.getName())) {
                    isIndex = true;
                    break;
                }
            }
            builder.append("\n    @Attribute\n");
            if (isIndex) {
                builder.append("    @Index\n");
            }

            if ((systemEntity.getPartition() != null) && systemEntity.getPartition().getName().equals(attribute.getName())) {
                builder.append("    @Partition\n");
            }
            builder.append("    ")
                   .append(attribute.getDataType())
                   .append(" ").append(attribute.getName()).append("\n");
        }

        builder.append("\n\n");

        for (final SystemRelationship relationship : systemEntity.getRelationships()) {

            String type;

            if ((relationship.getRelationshipType() == RelationshipType.ONE_TO_MANY.ordinal()) ||
                    (relationship.getRelationshipType() == RelationshipType.MANY_TO_MANY.ordinal())) {
                final String genericType = relationship.getInverseClass();
                final String collectionClass = List.class.getName();
                type = collectionClass + "<" + genericType + ">";
            } else {
                type = relationship.getInverseClass();
            }

            builder.append("    @Relationship(type = ")
                   .append(RelationshipType.values()[relationship.getRelationshipType()].name())
                   .append(", inverseClass = ")
                   .append(relationship.getInverseClass())
                   .append(", inverse = \"")
                   .append(relationship.getInverse())
                   .append("\", fetchPolicy = ")
                   .append(FetchPolicy.values()[relationship.getFetchPolicy()].name())
                   .append(", cascadePolicy = ")
                   .append(CascadePolicy.values()[relationship.getCascadePolicy()].name())
                   .append(", loadFactor = ")
                   .append(String.valueOf(relationship.getLoadFactor()))
                   .append(")\n    ")
                   .append(type)
                   .append(" ")
                   .append(relationship.getName())
                   .append("\n");

            // Get the base Descriptor so that we ensure they get generated also
            try {
                schemaContext.getBaseDescriptorForEntity(Class.forName(relationship.getInverseClass()));
            } catch (OnyxException | ClassNotFoundException ignore) {
            }
        }

        builder.append("\n}\n");

        writeClassContents(outputDirectory, systemEntity.getName(), builder.toString(), ".xtend");
    }

    /**
     * Generate Write a class to disk.  This is used for remote purposes.  Since we do not have a handle on the entity descriptors, we use the system entity to load
     *
     * @param systemEntity     System entity
     * @param databaseLocation Database location
     */
    @SuppressWarnings("ResultOfMethodCallIgnored")
    public synchronized static void writeClass(final SystemEntity systemEntity, final String databaseLocation, SchemaContext context) {
        if(context instanceof CacheSchemaContext)
            return;

        final String outputDirectory = databaseLocation + File.separator + SOURCE_ENTITIES_DIRECTORY;
        new File(outputDirectory).mkdirs();

        AtomicReference<String> idType = new AtomicReference<>();
        AtomicReference<String> idName = new AtomicReference<>();

        WRITTEN_CLASSES.add(systemEntity.getName());

        for (SystemAttribute attribute : systemEntity.getAttributes()) {
            if (attribute.getName().equals(systemEntity.getIdentifier().getName())) {
                idType.set(attribute.getDataType());
                idName.set(attribute.getName());
                break;
            }
        }

        writeXtendClass(systemEntity, databaseLocation, context, idType.get(), idName.get());
    }

    @SuppressWarnings({"WeakerAccess", "unused"})
    public EntityClassLoader() {
    }

    /**
     * Load the class source from file and load it into class loader.
     *
     * @param context Schema Context
     */
    @SuppressWarnings("ResultOfMethodCallIgnored")
    private static void loadClasses(final SchemaContext context, String location) {

        final File entitiesSourceDirectory = new File(location + File.separator + SOURCE_ENTITIES_DIRECTORY);
        final File entitiesGeneratedDirectory = new File(location + File.separator + GENERATED_ENTITIES_DIRECTORY);
        entitiesGeneratedDirectory.mkdirs();
        entitiesSourceDirectory.mkdirs();

        final JavaCompiler compiler = ToolProvider.getSystemJavaCompiler();
        final StandardJavaFileManager fileManager = compiler.getStandardFileManager(null, null, null);

        final List<File> classes = new ArrayList<>();

        try {
            forEachClass(new File(entitiesSourceDirectory.getPath()), classes::add);

            fileManager.setLocation(StandardLocation.CLASS_OUTPUT, Collections.singletonList(entitiesGeneratedDirectory));

            // Compile the file
            compiler.getTask(null, fileManager, null, null, null, fileManager.getJavaFileObjectsFromFiles(classes)).call();

            fileManager.close();

            final URLClassLoader systemClassLoader = (URLClassLoader) ClassLoader.getSystemClassLoader();

            addClassPaths(context);

            forEachClass(new File(entitiesSourceDirectory.getPath()), o -> {
                String path = o.getPath().replace(entitiesSourceDirectory.getPath() + File.separator, "");
                path = path.replaceAll("\\.java", "");
                path = path.replaceAll("\\\\", ".");
                path = path.replaceAll("/", ".");
                try {
                    LOADED_CLASSES.add(systemClassLoader.loadClass(path).getName());
                } catch (ClassNotFoundException e1) {
                    e1.printStackTrace();
                }
            });

        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public static void forEachClass(File root, Consumer<File> consumer) {

        File[] list = root.listFiles();

        for (File f : list) {
            if (!f.isDirectory() && !f.isHidden() && f.getPath().endsWith(".java")) {
                consumer.accept(f);
            }
            if (f.isDirectory()) {
                forEachClass(f, consumer);
            }
        }
    }

    /**
     * Load the class source from file and load it into class loader.
     *
     * @param context Schema Context
     */
    public static void loadClasses(final SchemaContext context) {
        loadClasses(context, context.getLocation());
    }

    @SuppressWarnings("unchecked")
    public static void addClassPaths(SchemaContext schemaContext) {
        final URLClassLoader systemClassLoader = (URLClassLoader) ClassLoader.getSystemClassLoader();

        // Add URL to class path
        final Class systemClass = URLClassLoader.class;

        try {
            final Method method = systemClass.getDeclaredMethod("addURL", URL.class);
            method.setAccessible(true);
            method.invoke(systemClassLoader, new File(schemaContext.getLocation() + File.separator + GENERATED_ENTITIES_DIRECTORY).toURI().toURL());
            method.invoke(systemClassLoader, new File(schemaContext.getLocation() + File.separator + GENERATED_QUERIES_DIRECTORY).toURI().toURL());
        } catch (Exception ignore) {
        }
    }

}

File: onyx-database/src/main/kotlin/com/onyx/classLoader/EntityClassLoader.kt
package com.onyx.classLoader

import com.onyx.entity.SystemEntity
import com.onyx.extension.common.ClassMetadata.classForName
import com.onyx.extension.common.catchAll
import com.onyx.persistence.annotations.values.CascadePolicy
import com.onyx.persistence.annotations.values.FetchPolicy
import com.onyx.persistence.annotations.values.IdentifierGenerator
import com.onyx.persistence.annotations.values.RelationshipType
import com.onyx.persistence.context.SchemaContext
import com.onyx.persistence.context.impl.CacheSchemaContext

import javax.tools.StandardLocation
import javax.tools.ToolProvider
import java.io.File
import java.io.FileWriter
import java.net.URL
import java.net.URLClassLoader
import java.util.*

/**
 * Created by timothy.osborn on 3/6/15.
 *
 *
 * This class saves the entity information and formats the source on disk
 */
//noinspection SpellCheckingInspection
object EntityClassLoader {

    private val LOADED_CLASSES: MutableSet<String> = HashSet()
    private val WRITTEN_CLASSES: MutableSet<String> = HashSet()

    private val GENERATED_DIRECTORY = "generated"
    private val GENERATED_ENTITIES_DIRECTORY = GENERATED_DIRECTORY + File.separator + "entities"
    private val GENERATED_QUERIES_DIRECTORY = GENERATED_DIRECTORY + File.separator + "queries"

    private val SOURCE_DIRECTORY = "source"
    private val XTEND_SOURCE_DIRECTORY = "xtend"
    private val SOURCE_ENTITIES_DIRECTORY = SOURCE_DIRECTORY + File.separator + "entities"
    private val XTEND_SOURCE_ENTITIES_DIRECTORY = XTEND_SOURCE_DIRECTORY + File.separator + "entities"

    /**
     * Write an enum to file
     * @param enumClass Enum Class Name
     * @param enumValues Enum values separated by comma and ended with a semicolon.
     * @param outputDirectory Source directory
     */
    private fun writeEnum(enumClass: String?, enumValues: String?, outputDirectory: String) {
        val nameTokens = enumClass!!.split("\\.".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        var packageName = "package "
        for (i in 0 until nameTokens.size - 1)
            packageName = packageName + nameTokens[i] + "."

        packageName = packageName.substring(0, packageName.length - 1)
        packageName += ";\n\n"

        val classDef = "public enum " + nameTokens[nameTokens.size - 1] + " { \n"
        val end = "\n}"
        val classContents = packageName + classDef + enumValues + end

        writeClassContents(outputDirectory, enumClass, classContents)
    }

    /**
     * Helper method used to write the class contents to a file
     *
     * @param outputDirectory Source directory
     * @param className Name of class including package
     * @param classContents Class source code
     * @param extension file extension
     */
    private fun writeClassContents(outputDirectory: String, className: String, classContents: String, extension: String = ".java") {
        // File output
        val classFile = File(outputDirectory + File.separator + className.replace("\\.".toRegex(), "/") + extension)
        classFile.parentFile.mkdirs()
        classFile.createNewFile()

        val file = FileWriter(outputDirectory + File.separator + className.replace("\\.".toRegex(), "/") + extension)

        file.write(classContents)
        file.flush()
        file.close()
    }

    /**
     * Write Xtend class in order to support DSL within the Database Browser
     *
     * @param systemEntity Class entity metadata
     * @param databaseLocation Location where the database lives
     * @param schemaContext Database context
     * @param idType Entity identifier type
     * @param idName Name of the identifier attribute
     */
    private fun writeXtendClass(systemEntity: SystemEntity, databaseLocation: String, schemaContext: SchemaContext, idType: String, idName: String) {
        val packageName = systemEntity.name.replace("." + systemEntity.className!!, "")
        val outputDirectory = databaseLocation + File.separator + XTEND_SOURCE_ENTITIES_DIRECTORY
        val className = systemEntity.className
        val generatorType = IdentifierGenerator.values()[systemEntity.identifier!!.generator.toInt()].toString()
        val fileName = systemEntity.fileName
        val idLoadFactor = systemEntity.identifier!!.loadFactor.toString()

        val builder = StringBuilder()
        builder.append("package ")
        builder.append(packageName)
        builder.append("\n" +
                "\n" +
                "import com.onyx.persistence.annotations.*\n" +
                "import com.onyx.persistence.*\n" +
                "import net.sagittarian.onyx.annotations.OnyxFields\n" +
                "import net.sagittarian.onyx.annotations.OnyxJoins\n" +
                "import org.eclipse.xtend.lib.annotations.Accessors" +
                "\n\n" +
                "@Entity(fileName = \"")
                .append(fileName)
                .append("\")\n")
                .append("@OnyxFields\n")
                .append("@OnyxJoins\n")
                .append("@Accessors\n")
                .append("class ")
                .append(className)
                .append(" extends ManagedEntity implements IManagedEntity\n{\n\n")
                .append("    @Attribute\n")
                .append("    @Identifier(generator = ")
                .append(generatorType)
                .append(", loadFactor = ")
                .append(idLoadFactor)
                .append(")\n    ")
                .append(idType)
                .append(" ")
                .append(idName)
                .append(";\n\n")

        systemEntity.attributes.forEach {
            if (it.name == systemEntity.identifier!!.name)
                return@forEach

            if (it.isEnum)
                writeEnum(it.dataType, it.enumValues, outputDirectory)

            val isIndex = systemEntity.indexes.firstOrNull { (name) -> name == it.name} != null

            builder.append("\n    @Attribute\n")
            if (isIndex)
                builder.append("    @Index\n")

            if (systemEntity.partition != null && systemEntity.partition!!.name == it.name)
                builder.append("    @Partition\n")

            builder.append("    ").append(it.dataType).append(" ").append(it.name).append("\n")
        }

        builder.append("\n\n")

        systemEntity.relationships.forEach {
            val type: String
            val genericType = it.inverseClass

            type = if (it.relationshipType.toInt() == RelationshipType.ONE_TO_MANY.ordinal || it.relationshipType.toInt() == RelationshipType.MANY_TO_MANY.ordinal) {
                val collectionClass = List::class.java.name
                "$collectionClass<$genericType>"
            } else {
                genericType
            }

            builder.append("    @Relationship(type = ")
                    .append(RelationshipType.values()[it.relationshipType.toInt()].name)
                    .append(", inverseClass = ")
                    .append(genericType)
                    .append(", inverse = \"")
                    .append(it.inverse)
                    .append("\", fetchPolicy = ")
                    .append(FetchPolicy.values()[it.fetchPolicy.toInt()].name)
                    .append(", cascadePolicy = ")
                    .append(CascadePolicy.values()[it.cascadePolicy.toInt()].name)
                    .append(", loadFactor = ")
                    .append(it.loadFactor.toString())
                    .append(")\n    ")
                    .append(type)
                    .append(" ")
                    .append(it.name)
                    .append("\n")

            // Get the base Descriptor so that we ensure they get generated also
            catchAll {
                schemaContext.getBaseDescriptorForEntity(classForName(genericType))
            }
        }

        builder.append("\n}\n")

        writeClassContents(outputDirectory, systemEntity.name, builder.toString(), ".xtend")
    }

    /**
     * Generate Write a class to disk.  This is used for remote purposes.  Since we do not have a handle on the entity descriptors, we use the system entity to load
     *
     * @param systemEntity     System entity
     * @param databaseLocation Database location
     */
    @Synchronized
    fun writeClass(systemEntity: SystemEntity, databaseLocation: String, context: SchemaContext) {
        if (context is CacheSchemaContext)
            return

        val outputDirectory = databaseLocation + File.separator + SOURCE_ENTITIES_DIRECTORY
        File(outputDirectory).mkdirs()

        WRITTEN_CLASSES.add(systemEntity.name)

        val attribute = systemEntity.attributes.find { it.name == systemEntity.identifier!!.name }!!
        writeXtendClass(systemEntity, databaseLocation, context, attribute.dataType!!, attribute.name)
    }

    /**
     * Load the class source from file and load it into class loader.
     *
     * @param context Schema Context
     * @param location Location to load classes from
     */
    @JvmOverloads
    fun loadClasses(context: SchemaContext, location: String = context.location) {

        val entitiesSourceDirectory = File(location + File.separator + SOURCE_ENTITIES_DIRECTORY)
        val entitiesGeneratedDirectory = File(location + File.separator + GENERATED_ENTITIES_DIRECTORY)
        entitiesGeneratedDirectory.mkdirs()
        entitiesSourceDirectory.mkdirs()

        val compiler = ToolProvider.getSystemJavaCompiler()
        val fileManager = compiler.getStandardFileManager(null, null, null)

        val classes = ArrayList<File>()

        File(entitiesGeneratedDirectory.path).walkTopDown().filter {it.isDirectory && !it.isHidden && it.path.endsWith(".java") }.forEach { classes.add(it) }
        fileManager.setLocation(StandardLocation.CLASS_OUTPUT, listOf(entitiesGeneratedDirectory))

        // Compile the file
        compiler.getTask(null, fileManager, null, null, null, fileManager.getJavaFileObjectsFromFiles(classes)).call()

        fileManager.close()

        val systemClassLoader = ClassLoader.getSystemClassLoader() as URLClassLoader

        addClassPaths(context)

        File(entitiesSourceDirectory.path).walkTopDown().filter {it.isDirectory && !it.isHidden && it.path.endsWith(".java") }.forEach {
            var path = it.path.replace(entitiesSourceDirectory.path + File.separator, "")
            path = path.replace("\\.java".toRegex(), "")
            path = path.replace("\\\\".toRegex(), ".")
            path = path.replace("/".toRegex(), ".")
            try {
                LOADED_CLASSES.add(systemClassLoader.loadClass(path).name)
            } catch (e:Exception) { e.printStackTrace() }
        }
    }

    /**
     * Add class paths for a database
     *
     * @param schemaContext Context of the database
     */
    private fun addClassPaths(schemaContext: SchemaContext) {
        val systemClassLoader = ClassLoader.getSystemClassLoader() as URLClassLoader

        // Add URL to class path
        val systemClass = URLClassLoader::class.java

        val method = systemClass.getDeclaredMethod("addURL", URL::class.java)
        method.isAccessible = true
        method.invoke(systemClassLoader, File(schemaContext.location + File.separator + GENERATED_ENTITIES_DIRECTORY).toURI().toURL())
        method.invoke(systemClassLoader, File(schemaContext.location + File.separator + GENERATED_QUERIES_DIRECTORY).toURI().toURL())
    }
}
